components/scream/src/share/io/tests/io.cpp

```diff
@@ -251,116 +264,265 @@ TEST_CASE("input_output_basic","io")
+      if (io_control.is_write_step(time)) {
+        output_stamps.push_back(time.to_string());
+        io_control.nsteps_since_last_write = 0;
+        io_control.timestamp_of_last_write = time;
+      }
     }
+
+    // Finalize the output manager (close files)
+    om.finalize();
   }
-  for (int jj=0;jj<num_levs;++jj) {
-    avg_val = (max_steps+1)/2.0*dt + (jj+1)/10.;  //note x0=(jj+1)/10 in this case.
-    REQUIRE(std::abs(f2_host(jj)-avg_val)<tol);
-  }
-  avg_input.finalize();
+
+  // Get a fresh new field manager
+  auto field_manager = get_test_fm(grid);
+  const auto& out_fields = field_manager->get_groups_info().at("output");
+  auto reset_fields = [&] () {
+    // Because the next test for input in this sequence relies on the same fields.  We set all field values
+    // to nan to ensure no false-positive tests if a field is simply not read in as input.
+    for (const auto& fname : out_fields->m_fields_names) {
+      auto f = field_manager->get_field(fname);
+      f.deep_copy(ekat::ScalarTraits<Real>::invalid());
+    }
+  };
   reset_fields();
 
-  // Check max output
-  // The max should be equivalent to the instantaneous because this function is monotonically increasing.
-  AtmosphereInput max_input(max_params,field_manager);
-  max_input.read_variables();
-  f1.sync_to_host();
-  f2.sync_to_host();
-  f3.sync_to_host();
-  for (int ii=0;ii<num_lcols;++ii) {
-    REQUIRE(std::abs(f1_host(ii)-(max_steps*dt+ii))<tol);
+  // We can use the produced output files to simultaneously check output quality and the
+  // ability to read input.
+  // NOTE: single-snap file start outputing at the final time only,
+  //       while multi-snap starts to output at the first time step
+  // First we re-construct the timestamp for the input file:
+  REQUIRE(output_stamps.size()>0);
+  auto input_params = get_in_params(output_type,output_freq_units, io_comm,output_stamps.front());
+  const Real tol = 1000*std::numeric_limits<Real>::epsilon();
+
+  // TODO: Create a small nc dummy file and a separate unit test which tests all input functions.
+  // Test that pio_inq_dimlen is correct, using a file from one of the above parameter lists.
+  {
+    auto test_filename = input_params.get<std::string>("Filename");
+    scorpio::register_file(test_filename,scorpio::Read);
+    Int test_gcols_len = scorpio::get_dimlen_c2f(test_filename.c_str(),"ncol");
+    REQUIRE(test_gcols_len==num_gcols);
+    scorpio::eam_pio_closefile(test_filename);
+  }
+
+  auto f1 = field_manager->get_field("field_1");
+  auto f2 = field_manager->get_field("field_2");
+  auto f3 = field_manager->get_field("field_3");
+  auto f4 = field_manager->get_field("field_packed");
+  auto f1_host = f1.get_view<Real*,Host>();
+  auto f2_host = f2.get_view<Real*,Host>();
+  auto f3_host = f3.get_view<Real**,Host>();
+  auto f4_host = f4.get_view<Real**,Host>();
+
+
+  if (output_type == "instant") {
+    // Check instant output
+    AtmosphereInput ins_input(input_params,field_manager);
```

Comment:
The creation of the reader, the call to `read_variables`, and the sync's to host can all be done once for all output types, before the if statement.